Retry publish crossgen as AOT (#65948)

agocke opened the pull request:
Trying to work through the problems with source build

agocke committed:
Publish crossgen as AOT if supported by the target platform (#65948)

Publishes crossgen as an AOT binary on Windows+Linux x64+ARM64, otherwise publishes as an R2R single file.

Closes #60016

(cherry picked from commit 0d1e04ba2a9ebc9d21a7f65db00407dd0056b547)

am11 commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -1,9 +1,48 @@
-    <OutputPath>$(RuntimeBinDir)/crossgen2</OutputPath>
-    <!-- The default value for macOS is false -->
-    <UseAppHost>true</UseAppHost>
-    <AppHostRuntimeIdentifier>$(Crossgen2PackageRID)</AppHostRuntimeIdentifier>
```
If this is not used anymore (this is the usage), please delete the definition as well https://github.com/dotnet/runtime/blob/bb5d94d3d6d91f4b1bbadcb5ff923ac97684e61b/Directory.Build.props#L186-L188

Note that source build is not the only aspect which brings complication in RID calculation, there are several other corner-cases, platform and configurations (portable=true/false, cross=true/false, tizen, musl, etc.) which break something and usually show up _after_ the changes are merged. 😅

agocke committed:
Set TargetFramework for the crossgen build to be NetCoreAppToolCurrent

agocke commented:
Which much help from @hoyosjs, @carlossanlop, and @krwq I think I have figured this out: the problem is that sometimes the crossgen2 package was restored as net6.0 and sometimes as net7.0. I can't identify from the logs *where* the net7.0 was coming from, but I think I can pass it down from the package build, which should hopefully ensure that it's always 6.0 at the time it matters.

agocke committed:
Remove crossgen2packagerid property

agocke committed:
Override in csproj

agocke committed:
Try deleting first

agocke committed:
Log the assets.json file

agocke committed:
Pass down DotNetBuildFromSource

agocke committed:
Up the message importance

agocke committed:
Try to skip restore in source build

agocke committed:
Typo

agocke committed:
Add PropertyGroup

agocke committed:
Try depending on restore

agocke committed:
Use PackageRID instead of RuntimeIdentifier

agocke committed:
Remove restore dependency

agocke committed:
Add log of RID

agocke committed:
Set RuntimeIdentifier in the project file

agocke committed:
Add print after restore

agocke committed:
only log while running publish

agocke committed:
Typo

agocke commented:
So, digging in more, I'm more confused about what's going on here. I wonder if there's something I don't understand about MSBuild static graph restore. 

@rainersigwald Any chance you can provide any insight here? The problem is that the `project.assets.json` does not seem to consistently get a RID included. It has `net6.0` instead of `net6.0/centos.7-x64`.

The build I'm having trouble with is at https://dev.azure.com/dnceng/public/_build/results?buildId=1709594&view=logs&j=000ba8e3-34f8-51d2-8d06-9b61f10256d0&t=ca9b6e26-8c16-57f9-673e-2bc0dc655f25

The important part of the log is

```
Microsoft.NETCore.App.Crossgen2 -> 
    Deleting project.assets.json
    Continuing with RID centos.7-x64
    Determining projects to restore...
    Restored /__w/1/s/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler.Diagnostics/ILCompiler.Diagnostics.csproj (in 270 ms).
    Restored /__w/1/s/artifacts/source-build/self/src/src/coreclr/tools/aot/crossgen2/crossgen2.csproj (in 269 ms).
    Restored /__w/1/s/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler.TypeSystem/ILCompiler.TypeSystem.csproj (in 269 ms).
    Restored /__w/1/s/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler.ReadyToRun/ILCompiler.ReadyToRun.csproj (in 269 ms).
    Restored /__w/1/s/artifacts/source-build/self/src/src/coreclr/tools/aot/ILCompiler.DependencyAnalysisFramework/ILCompiler.DependencyAnalysisFramework.csproj (in 29 ms).
    FileContents: /__w/1/s/artifacts/source-build/self/src/artifacts/obj/coreclr/crossgen2/project.assets.json
    {
    "version": 3,
    "targets": {
    "net6.0": {
    "Microsoft.Build.Tasks.Git/1.2.0-beta-22174-02": {
    "type": "package",
...
```

The important bits are

```
    Deleting project.assets.json
    Continuing with RID centos.7-x64
```

and
```
   "targets": {
    "net6.0": {
```

I don't understand why that is not `"targets": { "net6.0/centos.7-x64"`. The code here is basically a straightforward MSBuild invocation of the target project: 
https://github.com/dotnet/runtime/pull/67636/files#diff-115c40d5f6cdfb8ebea1490c13581e76b60c3430cbf13f45c270fc5ec383daffR36-R45

The RuntimeIdentifier should be a global property, and I don't understand why it's not factoring into the restore.

agocke commented:
also + @dsplaisted in case you have any ideas or can point me in the right direction

dsplaisted commented:
Even if a RuntimeIdentifier is specified, the assets file will still have a RID-less "targets" entry in it, and it will be the first one.  Later on in the file there should also be a target for "net6.0/centos.7-x64".

```
   "targets": {
    "net6.0": {
```

agocke commented:
Thanks Daniel, you're right, this file is weirdly mangled -- another RID might be in there somewhere, but there's a bunch of formatting errors, so I can't tell. I'll see if I can get more info.

agocke committed:
Escape characters

agocke committed:
Use cat instead of reading lines

agocke commented:
@dsplaisted @rainersigwald Not sure if you all are the right people to ask about this, happy to go to someone else if you have a recommendation.

agocke committed:
Try disabling assets cache

agocke committed:
Adjust RIDs

agocke committed:
Set RID to package RID for source build

agocke committed:
Typo

agocke committed:
Typo

agocke committed:
Adjust crossgen path on source build

agocke committed:
Fix crossgen2tool location

agocke committed:
Removing logging

agocke commented:
OK, I think this is ready for review. I can't say 100%, but I think the cause of this is a bug in the `ResolvePackageAssets` task.

The scenario is as follows: the runtime build off kicks off and restores all the projects. This includes `crossgen2.csproj`, and it starts by restoring with either no RID or the default RIDs. That crossgen is used in the bootstrap phase to compile System.Private.CoreLib for the test suite.

The second phase is creating the crossgen package. This is still part of the original MSBuild dispatch. Here we need to publish crossgen2 as self-contained/NativeAOT (depending on the platform), so I added code to invoke the MSBuild task on the crossgen2 project. For source-build, this involves using the native platform RID (like centos.7-x64).

When I made that change, this started failing about 50% of the time, with the error that `centos7-x64` didn't exist in the project.assets.json file. I've added a bunch of logging, and modified the build to delete the existing `project.assets.json` before building the crossgen package. No change.

I think the root of the problem is that there's some internal caching in this task that isn't reloading the file. I can't pretend to understand all that's going on here, but I wonder if the eviction policy for https://github.com/dotnet/sdk/blob/09b31215867d1ffe4955fd5b7cd91eb552d3632c/src/Tasks/Microsoft.NET.Build.Tasks/LockFileCache.cs#L39 might be somehow incorrect.

Regardless, I seem to have managed to solve this by always using the target RID during source build. Hopefully this will no longer be flaky

agocke committed:
Remove target reference

agocke committed:
Set PackageRID to include portable OS when building portable bits

am11 commented:
Change in the last commit 4dacc8f071f2a5421a9eecc0319f1702b21200d4 seems unrelated and it will not work for the case when we cross build for a non-Linux target OS on a Linux host. FreeBSD CI failure is related to that; as in the CI, we are cross building FreeBSD on Linux.

`PackageRID` is the RID used to restore nuget packages during the build and `OutputRid` is the RID of the (shipping) packages produced by the build. In case of cross build, we want `PackageRID` to match the host, not the target.

agocke commented:
Yup, I think I have a fix for that

dsplaisted commented:
@agocke Using the MSBuild task to restore and build a project is tricky, because the restore may download .props and .targets files, but MSBuild by default won't re-evaluate the project when running a different target in it if none of the global properties have changed.

I'm not sure if that's what's happening here, the fact that the build is using the MSBuild task to kick off restores and builds of projects is suspicious, but the symptoms where the assets file doesn't seem to have the right data doesn't seem like it would be caused by this.

To do this correctly, you need to specify a different global property in Restore as you do for Build in order to force a new evaluation.  Here's an example of how Arcade does this: https://github.com/dotnet/arcade/blob/884e6caa8158e6d1ccbac243bd7a83428ba63ffa/src/Microsoft.DotNet.Arcade.Sdk/tools/Build.proj#L187-L197

rainersigwald commented:
> To do this correctly, you need to specify a different global property in Restore as you do for Build in order to force a new evaluation.

That's not good enough; it works only if the files touched by restore don't exist on the first time through. If they exist but get updated there's no way to do this in tasks.

agocke committed:
Revert change to PackageRID and instead add musl check for _packageOS

agocke commented:
🙁 Yeah, I'm starting to see some of the associated problems.

Am I generally right in thinking that the best way to approach this is to start using separate builds, i.e. stop calling the MSBuild task and start shelling out to a proper `dotnet build`?

agocke commented:
In particular, this was incredibly difficult to track down, and I'm wondering if a blanket approach of favoring build isolation for individual assets is the right choice. Seems like it would at worst be slower and more clunky, but it would avoid very tricky bugs.

am11 commented on Directory.Build.props:
```diff
@@ -153,7 +153,8 @@
     <_runtimeOS Condition="'$(PortableBuild)' == 'true'">$(_portableOS)</_runtimeOS>
 
-    <_packageOS Condition="'$(CrossBuild)' == 'true'">$(_hostOS.ToLowerInvariant)</_packageOS>
+    <_packageLibc Condition="'$(_packageLibc)' == '' and $(_runtimeOS.Contains('musl'))">-musl</_packageLibc>
```
nit: we normally do not override private properties (prefixed with `_`) from command-line / import-site.
I think here we do not intend to override this value from command-line:
```suggestion
    <_packageLibc Condition="$(_runtimeOS.Contains('musl'))">-musl</_packageLibc>
```

agocke committed:
Update Directory.Build.props

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

agocke committed:
Change to use RuntimeIdentifiers instead of RuntimeIdentifier, and avoid changing crossgen path

agocke commented:
I think this is ready to bring back @dotnet/runtime-infrastructure. I've run 10 or so CI runs and there's no more flakiness in source build.

ViktorHofer commented on eng/Subsets.props:
```diff
@@ -71,8 +71,7 @@
                                         '$(BuildAllConfigurations)' == 'true'">libs.native+</DefaultLibrariesSubsets>
-    <DefaultLibrariesSubsets>$(DefaultLibrariesSubsets)libs.sfx+libs.oob</DefaultLibrariesSubsets>
-    <DefaultLibrariesSubsets Condition="'$(DotNetBuildFromSource)' != 'true'">$(DefaultLibrariesSubsets)+libs.pretest</DefaultLibrariesSubsets>
+    <DefaultLibrariesSubsets>$(DefaultLibrariesSubsets)libs.sfx+libs.oob+libs.pretest</DefaultLibrariesSubsets>
```
Merge conflict? You removed the libs.pretest subset.

ViktorHofer replied:
The pretest.proj and externals.proj files don't work during sourcebuild as they rely on prebuilds that aren't source buildable.

ViktorHofer commented on eng/liveBuilds.targets:
```diff
@@ -222,4 +223,18 @@
     <BundledRuntimeIdentifierGraphFile>$(RuntimeIdGraphDefinitionFile)</BundledRuntimeIdentifierGraphFile>
   </PropertyGroup>
+
+  <Target Name="RewriteRuntimePackDir"
+          Condition="'$(RunningPublish)' == 'true'"
+          DependsOnTargets="ResolveRuntimeFilesFromLocalBuild"
+          BeforeTargets="ResolveRuntimePackAssets">
+      <ItemGroup>
+        <!-- Remove AspNetCore runtime pack since we don't build it locally -->
+        <ResolvedRuntimePack Remove="Microsoft.AspNetCore.App.Runtime.$(RuntimeIdentifier)" />
+
+        <ResolvedRuntimePack Update="Microsoft.NETCore.App.Runtime.$(RuntimeIdentifier)">
+          <PackageDirectory>$(MicrosoftNetCoreAppRuntimePackDir)</PackageDirectory>
+        </ResolvedRuntimePack>
+      </ItemGroup>
+  </Target>
```
This logic (picking the local runtime package directory when resolving the runtime pack) is already handled by targetingpack.targets which is used in many places in the repo. Can't you use that instead?

agocke replied:
I had trouble with this initially, but it may be easier now. Let me get to that in a follow-up PR.

ViktorHofer replied:
Which projects need this target in your change? Can we move that into a file that isn't imported by everything in the repository? MSBuild (target) evaluation isn't for free.

agocke replied:
Good point -- I think it's just crossgen2. I'll try to move it there.

agocke committed:
Remove libs.pretest from sourcebuild set

agocke committed:
Move RewriteRuntimePackDir to crossgen2 project

ViktorHofer commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -46,4 +46,18 @@
+          DependsOnTargets="ResolveRuntimeFilesFromLocalBuild"
+          BeforeTargets="ResolveRuntimePackAssets">
+      <ItemGroup>
+        <!-- Remove AspNetCore runtime pack since we don't build it locally -->
```
Is removing the AspNetCore runtime pack actually necessary? I never hit issue with it being available as long as no projects use it.

ViktorHofer replied:
ping @agocke for that question

ViktorHofer replied:
@agocke did you see that question? :)

ViktorHofer commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -46,4 +46,18 @@
   <Import Project="$(CoreCLRBuildIntegrationDir)Microsoft.DotNet.ILCompiler.targets"
           Condition="'$(NativeAotSupported)' == 'true' and '$(RunningPublish)' == 'true'" />
 
+  <Target Name="RewriteRuntimePackDir"
```
Maybe add a TODO to consolidate this logic with targetingpack.targets and/or create a tracking issue?

agocke replied:
Yeah, I'll file a tracking issue.

agocke commented:
Ah, right, this is failing now in source build because `libs.pretest` is needed to create the runtimelist.xml file for the runtime pack, but it's not there in source build.

agocke committed:
Use live host in source build

agocke commented:
OK, I re-added libs.pretest, but I changed `externals` to always use the live host in source build, so that should eliminate the binary dependency concern.

agocke commented on src/libraries/externals.csproj:
```diff
@@ -9,7 +9,7 @@
     <BinPlaceRuntime>false</BinPlaceRuntime>
     <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
-    <UseLiveBuiltDotNetHost Condition="'$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6'">true</UseLiveBuiltDotNetHost>
+    <UseLiveBuiltDotNetHost Condition="'$(DotNetBuildFromSource)' == 'true' or '$(TargetArchitecture)' == 's390x' or '$(TargetArchitecture)' == 'armv6'">true</UseLiveBuiltDotNetHost>
```
Yup, that was my intent. I think long term we should force building the host in all cases. It's small and very cheap compared to building the CLR. The additional build complexity of pulling from an external package is not worth it, IMO.

am11 replied:
> we should force building the host in all cases.

I agree. Especially when singlefilehost builds as part of clr subset these days, it makes more sense (it will also help community supported platforms). Ideally we should dogfood live corehost components in all such cases (https://github.com/dotnet/runtime/pull/58952 proves that it is possible). 🙂

ViktorHofer replied:
> Yup, that was my intent. I think long term we should force building the host in all cases. It's small and very cheap compared to building the CLR. The additional build complexity of pulling from an external package is not worth it, IMO.

Absolutely agree. It just needs someone doing it :)

agocke replied:
It's on my personal backlog, I just keep getting distracted :)

agocke commented:
@ViktorHofer Anything else, or is this ready to go?

ViktorHofer commented on Directory.Build.props:
```diff
@@ -153,7 +153,8 @@
     <_runtimeOS Condition="'$(PortableBuild)' == 'true'">$(_portableOS)</_runtimeOS>
 
-    <_packageOS Condition="'$(CrossBuild)' == 'true'">$(_hostOS.ToLowerInvariant)</_packageOS>
+    <_packageLibc Condition="$(_runtimeOS.Contains('musl'))">-musl</_packageLibc>
+    <_packageOS Condition="'$(CrossBuild)' == 'true'">$(_hostOS.ToLowerInvariant)$(_packageLibc)</_packageOS>
```
@am11 just to double check, does that change look right to you?

am11 replied:
@agocke, could you tell what is this fixing?

IMO, we should run official build against the PR as we did previously so we can get some confidence from the package validation step.

ps - what I learned previously about this part of the configuration (`<PropertyGroup Label="CalculateOS">`) is that very tricky to be sure about this part of the configuration. And when everything was green (incl. outerloop and official builds), it broke Tizen, which was reported few days later. 😁

agocke replied:
Ah, yes, the problem is that PackageRID was `linux-x64` on alpine, when it should be `linux-musl-x64`.

ViktorHofer commented on src/coreclr/tools/aot/crossgen2/crossgen2.csproj:
```diff
@@ -1,9 +1,63 @@
-    <UseAppHost>true</UseAppHost>
-    <AppHostRuntimeIdentifier>$(Crossgen2PackageRID)</AppHostRuntimeIdentifier>
+    <OutputPath>$(RuntimeBinDir)crossgen2</OutputPath>
+    <DisablePackageAssetsCache Condition="'$(DotnetBuildFromSource)' == 'true'">true</DisablePackageAssetsCache>
```
Why is that setting required?

agocke replied:
Ah, I can probably remove that now. The problem was that we do two restores: one from `Build.proj`, and a second one here when we call Publish on `crossgen2`. There was a big problem with caching here.

I've now changed the RuntimeIdentifiers to include the source build RID in the initial restore, so even if it's cached, it should be OK. I'll remove that.

ViktorHofer reviewed (commented):
Just a few remaining questions 👍

agocke committed:
Remove package assets cache disabling

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -13,56 +14,70 @@
-      <ReferenceCopyLocalPaths Include="@(LibrariesRuntimeFiles)" TargetPath="tools/" />
-    </ItemGroup>
+  <Target Name="RunPublishedCrossgen" AfterTargets="PublishCrossgen"
+          Condition="'$(TargetOs)' == '$(HostOs)' and '$(TargetArchitecture)' == '$(BuildArchitecture)'">
```
Consistency :)

```suggestion
          Condition="'$(TargetOS)' == '$(HostOS)' and '$(TargetArchitecture)' == '$(BuildArchitecture)'">
```

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -13,56 +14,70 @@
+      <_TargetsToRun>Restore;Publish;PublishItemsOutputGroup</_TargetsToRun>
+    </PropertyGroup>
+
+    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
```
Why can't this be a ProjectReference instead? ProjectReferences allow to specify additional properties and the target(s) to invoke.

agocke replied:
I had no idea. I'd rather do that in a follow-up PR, since this is a very important deliverable for .NET 7 and I'd like to get dogfooding feedback as soon as possible.

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -13,56 +14,70 @@
+              ;RuntimeIdentifier=$(PackageRID)
+              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
+              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets
+              ;DotNetBuildFromSource=$(DotNetBuildFromSource)"
```
Isn't that property set globally? Global properties flow automatically into msbuild tasks or into other ProjectReferences.

agocke replied:
Yeah, I was pulling my hair out and trying to test that theory :)

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -13,56 +14,70 @@
+      <Output TaskParameter="ConsoleOutput" PropertyName="CrossgenOutput" />
+      <Output TaskParameter="ExitCode" PropertyName="CrossgenExitCode" />
+    </Exec>
+    <Error Text="Crossgen failed with code $(CrossgenExitCode), output: $(CrossgenOutput)" Condition="$(CrossgenExitCode) != 0" />
```
Why do you need to capture crossgen's output and re-emit it to the console? Doesn't crossgen already do that? When an msbuild Exec task fails, execution is aborted.

agocke committed:
Update src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

agocke committed:
PR comments

agocke committed:
Typo

agocke committed:
PR feedback

agocke commented:
@ViktorHofer Anything else high-priority? I've opened an issue for follow-up work.

agocke commented:
@jkoritzinsky Does this new strategy also look OK to you? I don't think much has changed except including the RID in the list of RuntimeIdentifiers in the first restore, so even if the assets.json is cached, it will still have compatible content.

ViktorHofer commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj:
```diff
@@ -13,56 +14,66 @@
-    <NativeRuntimeAsset Include="$(DotNetHostBinDir)/$(LibPrefix)hostfxr$(LibSuffix)" TargetPath="tools/" />
-    <NativeRuntimeAsset Include="$(DotNetHostBinDir)/$(LibPrefix)hostpolicy$(LibSuffix)" TargetPath="tools/" />
-  </ItemGroup>
+    <PropertyGroup>
+      <_TargetsToRun>Restore;Publish;PublishItemsOutputGroup</_TargetsToRun>
+    </PropertyGroup>
+
+    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
+             Properties="RunningPublish=true
+              ;RuntimeIdentifier=$(PackageRID)
+              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
+              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets"
+             Targets="$(_TargetsToRun)">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="_RawCrossgenPublishFiles" />
+    </MSBuild>
```
You really want to avoid invoking restore in the same task as other targets run.

The pattern that I followed over the years that worked very well is to have two MSBuild task invocations. One for restore and one for anything else. What you currently have might work but is probably fragile and might easily break in the future. I recommend to add another msbuild task before the one that you currently have:

```suggestion
<MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Restore"
             Properties="MSBuildRestoreSessionId=$([System.Guid]::NewGuid())
              ;RunningPublish=true
              ;RuntimeIdentifier=$(PackageRID)
              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets" />

    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Publish;PublishItemsOutputGroup"
             Properties="RunningPublish=true
              ;RuntimeIdentifier=$(PackageRID)
              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets">
      <Output TaskParameter="TargetOutputs"
              ItemName="_RawCrossgenPublishFiles" />
    </MSBuild>
```

agocke committed:
Update src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

agocke commented:
@ViktorHofer Done, anything else?

ViktorHofer commented:
I do have a lot more questions on why we need to restore that project separately outside of the upfront static graph repo restore but I don't think they should block this PR to going in.

Just wanted to make sure that you understand that a lazy just-in-time restore is the thing that eventually causes a build to fail in the middle. I saw such cases with the libraries linker tests which do something similar. But we can discuss that in your follow-up issue, if you want to :)

agocke commented:
Makes sense. First thing in the follow-up can be to remove the restore. It may no longer be necessary.
